rfcs/0136-stabilize-incrementally.md

```diff
@@ -0,0 +1,374 @@
+
+## Flakes are very popular
+
+And they will be stabilized, with minimal delay.
+
+Firstly, there is no feature work proposed in the interim --- the splitting of Nix is just partitioning existing functionality, with some behind-the-scenes refactors needed to make that possible.
+The earlier steps have been carefully designed to be either easier or already started ("de-risked") to do our best to convince those that are concerned about the current "limbo" that stabilization will be gotten to no matter what, even if the splitting fails.
+
+## Difficulties in the roll-out
+
+Process concerns are addressed by having a clear process, with clear outcomes, before any stabilization is begun.
+That is the purpose of this RFC!
+
+## Flakes are criticized for eating other Nix features
+
+We do *not* propose adding delay to once-again separate those features from Flakes, as that would introduce more feature work which would delay stabilization and be unacceptable to the pro-Flakes faction.
+
+That said, an ancillary incremental beneift of incremental stabilization is to bolster a *sense* of layering in Nix's user interface that has been, according to some people, lost.
+Ensuring that these two subsets of the new CLI --- without Flakes, and without Nix language support --- do in fact make sense in isolation will provide a "scaffolding" upon which interested parties can later introduce generalized features like search and pure eval without Flakes.
+
+The hope is that such scaffolding will assuage this faction their concerns are heard without holding things up.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main downside is a delay from the splitting Nix process, and then a delay between the stabilization steps.
+
+The fact that [RFC 134] is already mostly implemented, and in code review, is hopefully reason enough to believe it shouldn't take much longer.
+That maximum delay waiting for that to complete should be dwarfed by duration of time we've spent "in limbo" without a clear plan to move forward.
+We therefore think that is a small and reasonable price to pay for the benefit of community harmony.
+
+An addition, in the detailed design there is an escape hatch saying that blocking on the implementation of [RFC 134] can be reconsidered things do indeed take longer than foreseen.
+
+The delay of "auditing and refining" and shouldn't represent time "idle" from a stabilization perspective.
+As long as we are making progress stabilizing features and having healthy discussions, we don't see any problem.
+
+**Step 0** is also designed to take the pressure off these possible sources of delay, giving us *something* to work on that is not blocked on [RFC 134] or anything else.
+
+# Alternatives
+[alternatives]: #alternatives
+
+We could, of course, just "rip off the band-aid" and stabilize everything at once.
```

Comment:
One alternative could be to break flake under a new name or call this v1 and announce that v2 will have breaking changes. [semver](https://semver.org/) write

> increment the MAJOR version when you make incompatible API changes

This describes could apply to the current usecase as it's what major versions are there for.

Reply:
Yeah the issue right now is not *having* multiple versions of Flakes, it's *maintaining* them. When there are not clear layer boundaries (bits of Flakes stuff all around `libexpr` and `libcmd`), the maintainence cost of 1 Flakes, let alone 2 Flakes(!), is much higher.

If we do succeed in make Flakes a separate project (my dream, not proposed as part of this RFC as written). Then yes, now the maintainence cost is much better. Out-of-tree front-ends can do whatever they want!